### ISSUE TITLE
[Deliverable] NAT traversal capabilities

### ISSUE THREAD

**Issue by gmega** [logos-storage/logos-storage-pm#5]
Using the Autonat support already present from logos-storage/logos-storage-nim#84, enable circuit relay v2. This may be tricky to actually test, some initial work to setup a NATted environment using docker and iptables [can be found here](https://github.com/gmega/codex-nat-env/).

**Comment by emizzle** [comment]
Useful resources:
https://docs.libp2p.io/concepts/nat/hole-punching/
https://docs.libp2p.io/concepts/nat/circuit-relay/

**Comment by emizzle** [comment]
We need to investigate how Waku uses Discv5 + circuit relay, ie https://github.com/waku-org/nwaku/blob/master/waku/factory/waku.nim#L99-L109

**Comment by AuHau** [comment]
[`HPService`](https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/services/hpservice.nim#L27) from `nim-libp2p` facilitates Hole Punching. 

Example how to integrate it - hole-pounching interop tests:  https://github.com/vacp2p/nim-libp2p/blob/master/interop/hole-punching/hole_punching.nim

**Comment by AuHau** [comment]
There is handy `nim-libp2p` service for managing the Circuit Relay for the client side: [`AutoRelayService`](https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/services/autorelayservice.nim#L21).

It should handle things like:
- relay discovery 
- reservation management - creation and renewal

More info:
https://deepwiki.com/search/i-am-interested-in-the-circuit_194ce25a-62ef-4cc6-b3d3-8583e93216c6

**Comment by emizzle** [comment]
Shouldn't we aim for DCUtR (eg [via `DcutrClient`](https://github.com/vacp2p/nim-libp2p/blob/061195195bb7496682a9205ee8ab2a8e6dd255dd/libp2p/services/hpservice.nim#L84C23-L90)), because imo that is _actual_ holepunching without a relay. My concern with relays is that they could be a huge bottleneck for throughput if nodes are downloading 10GB message archive files.

**Comment by AuHau** [comment]
> Shouldn't we aim for DCUtR (eg [via `DcutrClient`](https://github.com/vacp2p/nim-libp2p/blob/061195195bb7496682a9205ee8ab2a8e6dd255dd/libp2p/services/hpservice.nim#L84C23-L90)), because imo that is _actual_ holepunching without a relay. My concern with relays is that they could be a huge bottleneck for throughput if nodes are downloading 10GB message archive files.

No ;-) You can't have hole-punching without a relay or some other centralized coordination point (like STUN server).

**Comment by AuHau** [comment]
Giuliano's writeups about his research into this topic:

- https://hackmd.io/UxYaQ3dqS1CUU8pxO6-PuQ
- https://hackmd.io/p6fL4TEkTgGljTBmWfXicw

**Comment by emizzle** [comment]
> No ;-) You can't have hole-punching without a relay or some other centralized coordination point (like STUN server).

Yes, I'm aware. I should have worded that differently. We need a relay at the minimum, but we should aim for DCUtR if we can to avoid the bottleneck of relaying traffic with a lot of data transfer. However, if two nodes can only connect via relay, then at least they can still connect, even if there is latency and limited throughput.

**Comment by AuHau** [comment]
Yes, that is the plan. Try hole-punching and fallback to relay. This is facilitated in the `HPService`, which, upon a new peer being added, will try to ensure there is at least one direct (non-relayed) connection to it. If there is none, it will try to create one using the DCUtR. But the probability of hole-punching is apparently not so big. 

Btw. was looking into some resources if this was measured in some study, and while I did not find concrete numbers (I didn't search that hard though), one interesting note I found was that apparently UDP punching has a higher chance of success compared to TCP one.

**Comment by emizzle** [comment]
> Btw. was looking into some resources if this was measured in some study, and while I did not find concrete numbers (I didn't search that hard though), one interesting note I found was that apparently UDP punching has a higher chance of success compared to TCP one.

That's very interesting. I guess the down side is browser support, but i wonder if there's a way we could fallback to TCP somehow. Definitely not an immediate problem to solve.

**Comment by AuHau** [comment]
Honestly, I would not stress about browser support really. IMHO, this is big step to decide if we want to support it **natively**. Look at how `js-ipfs` and now `helia` "works". It really never performed well in browser settings. Most probably, the way to do "browser support" is using some light-client mechanisms, where most if not all "p2p-ing" will be delegated to some full client.

**Comment by emizzle** [comment]
That makes sense, but I was under the impression that light clients still needed access to the DHT, hence the issue with UDP and why libp2p Kademlia DHT used TCP.

**Comment by AuHau** [comment]
The AutoNat service has callback set using `statusAndConfidenceHandler()` which is called when the reachability of the node observed by AutoNat changes. This should be the driving routine that will guide switching the Discv5's client/server mode.

**Comment by AuHau** [comment]
Another thing to note is the default values for the Circuit Relays limits: 

https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/protocols/connectivity/relay/relay.nim#L42-L47

The relays are limited to two hours (with the default value being one hour) reservation. Maximum of 128 circuits per node and most importantly, a data limit of **131 072 bytes** (I think?). This will need to be tweaked for our use case; otherwise the unreachable nodes will be just spamming reservations as they will be trying to download some bigger files.

**Comment by AuHau** [comment]
Looking on `go-ipfs (kubo)` defaults, they are not changing them, but allow users to configure those: https://deepwiki.com/search/does-the-project-changes-the-d_9308f507-25c2-4660-9a50-4aab62cb0337

**Comment by emizzle** [comment]
> data limit of 131 072 bytes (I think?). This will need to be tweaked for our use case; otherwise the unreachable nodes will be just spamming reservations as they will be trying to download some bigger files.

Kubo utilises a max reservations per ip and ASN, which may be useful here. 

Are you thinking that we'd need to massively [increase the limits](https://deepwiki.com/search/what-is-an-asn-in-regards-to-c_f307c973-10ac-4ff9-84c7-9faa69261e75)? Also interesting is that apparently v1 is a better choice for larger data transfers than v2.

**Comment by AuHau** [comment]
Well, we will have to see about the limits. Let's see how it behaves first. But generally, I have just a hunch that for transferring hundreds of MB, this will have too large overheads. But then this could be worked around by exposing this limits as Kubo does and run our own fleet of relays that will have buffed these limits. Potentially there could be prioritization of the "large limit" relays which could go to the custom `AutoRelayService` implementation.

> Also interesting is that apparently v1 is a better choice for larger data transfers than v2.

What do you mean? AutoNAT?

**Comment by 2-towns** [comment]
Implementation plan: https://hackmd.io/3DR2Of1qS9iwXl_qdGdkKg
